### PR TITLE
[bitnami/cassandra] Add Config Parameter for Host Network and HostPorts

### PR DESCRIPTION
**Pull request by jmederer**
**Description of the change**
Now you can use (if needed) cassandra with hostNetwork.
And you can easily configure all the Ports of the cassandra Container.

**Benefits**
You can now connect your K8s cassandra cluster with any external cassandra cluster. 
This is really helpful if you're onPremise and got trouble with the "Broadcast_adress". 
HostNetwork maybe the last Change to run it.

**Possible drawbacks**

**Applicable issues**
The Downside is till now - the metrics (cassandra-exporter) Container won't accept any Config-Parameter (except a Config-Map would be an approach).
therefore this Container is not ready now for Host Ports.

  

**Additional information**

**Checklist** 
- [X] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [X] Variables are documented in the README.md
- [X] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)

:warning: Keep in mind that if you want to make changes to the kubeapps chart, please implement them in the [kubeapps repository](https://github.com/kubeapps/kubeapps/tree/master/chart/kubeapps). This is only a synchronized mirror.

### CODE REVIEW

File: bitnami/cassandra/Chart.yaml [hunk 1]
```diff
@@ -23,4 +23,4 @@ name: cassandra
 sources:
   - https://github.com/bitnami/bitnami-docker-cassandra
   - http://cassandra.apache.org
-version: 7.1.6
+version: 7.1.7
```
Comment (juan131) [inline bitnami/cassandra/Chart.yaml:26]: You're adding a new functionality. So please, bump the minor version instead.

File: bitnami/cassandra/README.md [hunk 2]
```diff
@@ -144,17 +144,25 @@ The following table lists the configurable parameters of the Cassandra chart and
...
+| `host.ports.cassandra.jmxPort`         | JMX Port on the Host and Container                                                                                      | `7199`                                                       |
+| `host.ports.cassandra.cqlPort`         | CQL Port on the Host and Container                                                                                      | `9042`                                                       |
+| `host.ports.cassandra.thriftPort`      | Thrift Port on the Host and Container                                                                                   | `9160`                                                       |
+| `host.ports.metrics.httpPort`          | HTTP Port on the Host and Container                                                                                     | `8080`                                                       |
+| `host.ports.metrics.jmxPort`           | JMX Port on the Host and Container                                                                                      | `5555`                                                       |
```
Comment (juan131) [inline bitnami/cassandra/README.md:165]: Can we use a single `hostNetwork` boolean parameter and rely on the default container ports? This is what we did in other charts (e.g. node-exporter)?
If you consider essential to customize the container ports, add a `containerPorts` array parameter as we do in other charts, e.g. see:
- https://github.com/bitnami/charts/blob/master/bitnami/keycloak/values.yaml#L200

File: bitnami/cassandra/templates/statefulset.yaml [hunk 3]
```diff
@@ -192,6 +196,18 @@ spec:
...
+              value: {{ .Values.host.ports.cassandra.intraPort | quote }}
+            {{- end }}
+            {{- if kindIs "valid" .Values.host.ports.cassandra.jmxPort }}
+            - name: CASSANDRA_JMX_PORT_NUMBER
+              value: {{ .Values.host.ports.cassandra.jmxPort | quote }}
+            {{- end }}
+            {{- if kindIs "valid" .Values.host.ports.cassandra.cqlPort }}
+            - name: CASSANDRA_CQL_PORT_NUMBER
+              value: {{ .Values.host.ports.cassandra.cqlPort | quote }}
+            {{- end }}
```
Comment (juan131) [inline bitnami/cassandra/templates/statefulset.yaml:210]: As I mentioned before, this should be based on a `containerPorts` array parameter (for consistency with other charts).

File: bitnami/cassandra/templates/statefulset.yaml [hunk 4]
```diff
@@ -247,16 +263,31 @@ spec:
...
+              {{- if .Values.host.hostNetwork }}
+              hostPort: {{ .Values.host.ports.cassandra.cqlPort }}
+              {{- end }}
             {{- if .Values.cluster.enableRPC }}
             - name: thrift
-              containerPort: 9160
+              containerPort: {{ .Values.host.ports.cassandra.thriftPort | default "9160" }}
+              {{- if .Values.host.hostNetwork }}
+              hostPort: {{ .Values.host.ports.cassandra.thriftPort }}
+              {{- end }}
```
Comment (juan131) [inline bitnami/cassandra/templates/statefulset.yaml:290]: Same here

**Review by juan131 — CHANGES_REQUESTED** [review]
Thanks so much for your contribution!
We really appreciate it! Please take a look to my suggestions when you have a chance

File: bitnami/cassandra/README.md (hunk 2 above)
Comment (jmederer) [reply bitnami/cassandra/README.md]: Perfect, i changed it to containerPorts.cassandra.*portname*

File: bitnami/cassandra/values.yaml [hunk 5]
```diff
@@ -54,6 +54,24 @@ commonAnnotations: {}
...
+containerPorts:
+  cassandra:
+    intra: 7000
+    tls: 7001
+    jmx: 7199
+    cql: 9042
+    thrift: 9160
+  metrics:
+    http: 8080
+    jmx: 5555
```
Comment (juan131) [inline bitnami/cassandra/values.yaml:69]: Please organize container ports as shown below:

```yaml
## Cassandra container ports to open
## If hostNetwork true: the hostPort is set identical to the containerPort
##
containerPorts:
  intra: 7000
  tls: 7001
  jmx: 7199
  cql: 9042
  thrift: 9160

(...)

metrics:

  (...)
  ## Cassandra metrics exporter container ports to open
  ## If hostNetwork true: the hostPort is set identical to the containerPort
  ##
  containerPorts:
    http: 8080
    jmx: 5555
```

Two reasons to do so:
- Consistency with other charts
- The rest of parameters affecting the metrics exporter are also organized under the `metrics` object.

File: bitnami/cassandra/README.md [hunk 6]
```diff
@@ -142,19 +142,27 @@ The following table lists the configurable parameters of the Cassandra chart and
...
+| `containerPorts.cassandra.jmx`         | JMX Port on the Host and Container                                                                                      | `7199`                                                       |
+| `containerPorts.cassandra.cql`         | CQL Port on the Host and Container                                                                                      | `9042`                                                       |
+| `containerPorts.cassandra.thrift`      | Thrift Port on the Host and Container                                                                                   | `9160`                                                       |
+| `containerPorts.metrics.http`          | HTTP Port on the Host and Container                                                                                     | `8080`                                                       |
+| `containerPorts.metrics.jmx`           | JMX Port on the Host and Container                                                                                      | `5555`                                                       |
```
Comment (juan131) [inline bitnami/cassandra/README.md:165]: I'd move all these `containerPorts` parameters under the "Statefulset parameters" section.

**Review by juan131 — CHANGES_REQUESTED** [review]
Thanks for implementing the changes! One last suggestion

File: bitnami/cassandra/README.md (hunk 6 above)
Comment (jmederer) [reply bitnami/cassandra/README.md]: Great, many thanks for your support.

File: bitnami/cassandra/templates/statefulset.yaml [hunk 7]
```diff
@@ -290,12 +321,25 @@ spec:
         - name: metrics
           image: {{ include "cassandra.metrics.image" . }}
           imagePullPolicy: {{ .Values.metrics.image.pullPolicy | quote }}
+          {{- if .Values.hostNetwork }}
+          env:
+            - name: CASSANDRA_EXPORTER_CONFIG_listenPort
+              value: {{ .Values.metrics.containerPorts.http | quote }}
+            - name:  CASSANDRA_EXPORTER_CONFIG_host
+              value: localhost:{{ .Values.containerPorts.cassandra.jmx }}
```
Comment (juan131) [inline bitnami/cassandra/templates/statefulset.yaml:329]: Just curious: why this weird "FIRST UPPERCASE, then lowercase" syntax for the environment variables?

File: bitnami/cassandra/Chart.yaml [hunk 8]
```diff
@@ -23,4 +23,4 @@ name: cassandra
 sources:
   - https://github.com/bitnami/bitnami-docker-cassandra
   - http://cassandra.apache.org
-version: 7.2.0
+version: 7.2.1
```
Comment (juan131) [inline bitnami/cassandra/Chart.yaml:26]: Remember to bump the **minor** version:
```suggestion
version: 7.3.0
```

**Review by juan131 — COMMENTED** [review]
Thanks so much for implementing the suggestions!
Once you fix the issue in the version, we're ready to merge it.

File: bitnami/cassandra/templates/statefulset.yaml (hunk 7 above)
Comment (jmederer) [reply bitnami/cassandra/templates/statefulset.yaml:329]: Yeah .-)
This in "preperation" for the `cassandra-exporter` because bitnami references the https://github.com/criteo/cassandra_exporter in https://github.com/bitnami/bitnami-docker-cassandra-exporter

An criteo gives actual the ability to set the config over this "UPPER lower" Case-Thing ->
https://github.com/criteo/cassandra_exporter/blob/master/docker/run.sh

Maybe a question from my side: Can we also start the bitnami/metric Container with a "launch" Script which is able to edit exporter.config? 

TL;DR
Both variables CASSANDRA_EXPORTER_CONFIG_host and CASSANDRA_EXPORTER_CONFIG_listenPort are only necessary for HOST-Network=true and set also the critord/cassandra-exporter image instead of the bitname/cassandra-exporter

Comment (juan131) [reply bitnami/cassandra/templates/statefulset.yaml:329]: > An criteo gives actual the ability to set the config over this "UPPER lower" Case-Thing ->
https://github.com/criteo/cassandra_exporter/blob/master/docker/run.sh

That's fair enough! It's "weird" they decided to use that syntax but it's not a major issue.

> Maybe a question from my side: Can we also start the bitnami/metric Container with a "launch" Script which is able to edit exporter.config?

I don't know if it's worth to do that for a simple container such an exporter.

**Review by juan131 — APPROVED** [review]
LGTM
